src/devices/cpu/mipsx/mipsxdasm.cpp

```diff
@@ -0,0 +1,588 @@
+// license:BSD-3-Clause
+// copyright-holders:David Haywood
+
+#include "emu.h"
+#include "mipsxdasm.h"
+
+u32 mipsx_disassembler::opcode_alignment() const
+{
+	return 4;
+}
+
+int mipsx_disassembler::get_ty(u32 opcode)
+{
+	return BIT(opcode, 30, 2);
+}
+
+int mipsx_disassembler::get_op(u32 opcode)
+{
+	return BIT(opcode, 27, 3);
+}
+
+int mipsx_disassembler::get_src1(u32 opcode)
+{
+	return BIT(opcode, 22, 5);
+}
+
+int mipsx_disassembler::get_src2_dest(u32 opcode)
+{
+	return BIT(opcode, 17, 5);
+}
+
+int mipsx_disassembler::get_compute_dest(u32 opcode)
+{
+	return BIT(opcode, 12, 5);
+}
+
+int mipsx_disassembler::get_compute_compfunc(u32 opcode)
+{
+	return BIT(opcode, 0, 12);
+}
+
+int mipsx_disassembler::get_offset(u32 opcode)
+{
+	return BIT(opcode, 0, 17);
+}
+
+int mipsx_disassembler::get_imm17(u32 opcode)
+{
+	return BIT(opcode, 0, 17);
+}
+
+int mipsx_disassembler::get_sq(u32 opcode)
+{
+	return BIT(opcode, 16, 1);
+}
+
+int mipsx_disassembler::get_disp(u32 opcode)
+{
+	return BIT(opcode, 0, 16);
+}
+
+int mipsx_disassembler::get_asr_amount(int shift)
+{
+	// TODO
+	return shift;
+}
+
+int mipsx_disassembler::get_sh_amount(int shift)
+{
+	// TODO
+	return shift;
+}
+
+
+std::string mipsx_disassembler::get_regname(u8 reg)
+{
+	// general purpose register 0 just acts as a constant 0, it can't be changed.
+	if ((reg == 0) && (SHOW_R0_AS_0))
+		return "#0";
+
+	const std::string regnames[32] = { "r0",  "r1",  "r2",  "r3",  "r4",  "r5",  "r6",  "r7",  "r8",  "r9",  "r10", "r11", "r12", "r13", "r14", "r15"
+									   "r16", "r17", "r18", "r19", "r20", "r21", "r22", "r23", "r24", "r25", "r26", "r27", "r28", "r29", "r30", "r31" };
+
+	return regnames[reg];
+}
+
+offs_t mipsx_disassembler::disassemble(std::ostream& stream, offs_t pc, const data_buffer& opcodes, const data_buffer& params)
+{
+	const u32 opcode = opcodes.r32(pc);
+	const u8 ty = get_ty(opcode);
+
+	switch (ty)
+	{
+	case 0:
+	{
+		const u8 op = get_op(opcode);
+		const int disp = util::sext(opcode, 16) * 4;
+		const u32 basepc = pc + 8;
+		const int src1 = get_src1(opcode);
+		const int src2 = get_src2_dest(opcode);
+		const std::string squash = get_sq(opcode) ? "sq" : "";
+
+		switch (op)
+		{
+		case 1:
+		{
+			if ((src1 == 0) && (src2 == 0))
+			{
+				// beq #0, #0, offset is used as an alias for unconditional branch
+				util::stream_format(stream, "bra%s %08x", squash, basepc + disp);
+			}
+			else
+			{
+				util::stream_format(stream, "beq%s %s, %s, %08x", squash, get_regname(src1), get_regname(src2), basepc + disp);
```

Comment:
Apologies for missing in the first review, but spaces between operands are not canonical.